File: riveredge-backend/src/apps/kuaizhizao/models/fai_order.py
"""FAI 首件检验单"""

from tortoise import fields

from core.models.base import BaseModel


class FaiOrder(BaseModel):
    class Meta:
        table = "apps_kuaizhizao_fai_orders"
        table_description = "快格轻制造 - FAI 首件检验单"
        indexes = [
            ("tenant_id",),
            ("fai_code",),
            ("status",),
            ("material_id",),
            ("work_order_id",),
        ]
        unique_together = [("tenant_id", "fai_code")]

    id = fields.IntField(pk=True, description="主键ID")
    fai_code = fields.CharField(max_length=50, description="FAI 编码")
    title = fields.CharField(max_length=200, description="标题")
    trigger_reason = fields.CharField(
        max_length=30, default="new_part", description="new_part/ecn/changeover/restart/customer"
    )
    status = fields.CharField(max_length=20, default="draft", description="单据状态")
    conclusion = fields.CharField(max_length=20, default="pending", description="pending/pass/fail")

    material_id = fields.IntField(null=True, description="物料ID")
    material_code = fields.CharField(max_length=50, null=True, description="物料编码")
    material_name = fields.CharField(max_length=200, null=True, description="物料名称")
    drawing_no = fields.CharField(max_length=100, null=True, description="图纸号")
    drawing_revision = fields.CharField(max_length=50, null=True, description="图纸版本")
    work_order_id = fields.IntField(null=True, description="工单ID")
    work_order_code = fields.CharField(max_length=50, null=True, description="工单号")
    inspection_plan_id = fields.IntField(null=True, description="来源质检方案ID")
    inspection_plan_code = fields.CharField(max_length=50, null=True, description="来源质检方案编码")

    # Form1 / Form2 摘要
    part_number = fields.CharField(max_length=100, null=True, description="零件号")
    part_name = fields.CharField(max_length=200, null=True, description="零件名称")
    serial_number = fields.CharField(max_length=100, null=True, description="序列号")
    lot_number = fields.CharField(max_length=100, null=True, description="批次号")
    material_spec = fields.CharField(max_length=200, null=True, description="材料规格")
    process_spec = fields.TextField(null=True, description="工艺/特殊过程摘要")
    organization_name = fields.CharField(max_length=200, null=True, description="组织/供应商名称")

    sample_size = fields.IntField(default=1, description="样本件数（Cp/Cpk）")
    cpk_summary = fields.JSONField(null=True, description="Cp/Cpk 摘要")
    drawing_file_url = fields.CharField(max_length=500, null=True, description="图纸文件")
    balloon_candidates = fields.JSONField(null=True, description="OCR/气泡候选（待确认）")
    attachments = fields.JSONField(null=True, description="附件")
    remarks = fields.TextField(null=True, description="备注")

    submitted_at = fields.DatetimeField(null=True, description="提交时间")
    approved_at = fields.DatetimeField(null=True, description="批准时间")
    approved_by = fields.IntField(null=True, description="批准人")
    approved_by_name = fields.CharField(max_length=100, null=True, description="批准人姓名")
    deleted_at = fields.DatetimeField(null=True, description="软删除")

File: riveredge-frontend/src/apps/kuaizhizao/services/fai-order.ts
import { apiRequest } from '../../../services/api';

export interface FaiCharacteristic {
  id?: number;
  sequence?: number;
  balloon_no?: string;
  characteristic_name: string;
  nominal_value?: number | null;
  upper_tolerance?: number | null;
  lower_tolerance?: number | null;
  unit?: string;
  measured_value?: number | null;
  sample_values?: number[];
  judgment?: string;
  gauge_id?: number;
  gauge_code?: string;
  gauge_name?: string;
  source_step_key?: string;
  remarks?: string;
}

export interface FaiBalloonCandidate {
  id?: string;
  balloon_no?: string;
  characteristic_name: string;
  nominal_value?: number | null;
  upper_tolerance?: number | null;
  lower_tolerance?: number | null;
  unit?: string;
  remarks?: string;
  /** 气泡中心，相对图纸 0~1 */
  x?: number;
  y?: number;
  anchor_x?: number;
  anchor_y?: number;
  source?: 'manual' | 'ocr';
}

export interface FaiBalloonOcrResult {
  candidates: FaiBalloonCandidate[];
  confidence_notes?: string | null;
}

export interface FaiOrder {
  id: number;
  uuid?: string;
  fai_code: string;
  title: string;
  trigger_reason: string;
  status: string;
  conclusion: string;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  drawing_no?: string;
  drawing_revision?: string;
  work_order_id?: number;
  work_order_code?: string;
  inspection_plan_id?: number;
  inspection_plan_code?: string;
  part_number?: string;
  part_name?: string;
  serial_number?: string;
  lot_number?: string;
  material_spec?: string;
  process_spec?: string;
  organization_name?: string;
  sample_size?: number;
  cpk_summary?: any;
  drawing_file_url?: string;
  balloon_candidates?: FaiBalloonCandidate[];
  attachments?: any;
  remarks?: string;
  characteristics?: FaiCharacteristic[];
  created_by?: number;
  created_by_name?: string;
  updated_by?: number;
  updated_by_name?: string;
  created_at?: string;
  updated_at?: string;
}

export interface FaiFairExport {
  fai_code: string;
  form1: Record<string, unknown>;
  form2: Record<string, unknown>;
  form3: Array<Record<string, unknown>>;
  cpk_summary?: any;
}

type ListEnvelope<T> = { items: T[]; total: number };

export const faiOrderApi = {
  list: (params?: Record<string, unknown>) =>
    apiRequest<ListEnvelope<FaiOrder>>('/apps/kuaizhizao/fai-orders', { method: 'GET', params }),
  get: (id: number) => apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}`, { method: 'GET' }),
  create: (data: Partial<FaiOrder> & { characteristics?: FaiCharacteristic[] }) =>
    apiRequest<FaiOrder>('/apps/kuaizhizao/fai-orders', { method: 'POST', data }),
  update: (id: number, data: Partial<FaiOrder> & { characteristics?: FaiCharacteristic[] }) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}`, { method: 'PUT', data }),
  delete: (id: number) => apiRequest(`/apps/kuaizhizao/fai-orders/${id}`, { method: 'DELETE' }),
  submit: (id: number) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/submit`, { method: 'POST' }),
  approve: (id: number) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/approve`, { method: 'POST' }),
  reject: (id: number, remarks?: string) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/reject`, {
      method: 'POST',
      data: { remarks },
    }),
  close: (id: number) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/close`, { method: 'POST' }),
  importFromPlan: (id: number, inspection_plan_id: number) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/import-from-plan`, {
      method: 'POST',
      data: { inspection_plan_id },
    }),
  fairExport: (id: number) =>
    apiRequest<FaiFairExport>(`/apps/kuaizhizao/fai-orders/${id}/fair-export`, { method: 'GET' }),
  saveBalloonCandidates: (id: number, candidates: FaiBalloonCandidate[], drawing_file_url?: string) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/balloon-candidates`, {
      method: 'PUT',
      data: { candidates, drawing_file_url },
    }),
  confirmBalloons: (id: number, candidates: FaiBalloonCandidate[], replace_existing = true) =>
    apiRequest<FaiOrder>(`/apps/kuaizhizao/fai-orders/${id}/confirm-balloons`, {
      method: 'POST',
      data: { candidates, replace_existing },
    }),
  balloonOcr: (id: number, file: Blob | File, persist = true) => {
    const formData = new FormData();
    const name = file instanceof File ? file.name : 'drawing.png';
    formData.append('file', file, name);
    return apiRequest<FaiBalloonOcrResult>(
      `/apps/kuaizhizao/fai-orders/${id}/balloon-ocr?persist=${persist ? 'true' : 'false'}`,
      {
        method: 'POST',
        body: formData,
      },
    );
  },
};
